### ISSUE TITLE
Burnout / neuron death

### ISSUE THREAD

**Issue by orcinus** [dvruette/sd-webui-fabric#34]
I seem to be getting horribly burnt out images no matter what i do.
I've tried eliminating all extensions, i tried custom (non anime) and stock VAE, i'm not using any hypernetworks or LoRAs.
I tried with my own fine-tuned (generalist) models as well as stock SD1.5 and 2.0.

I first thought that using too many upvote/downvote images causes burnout, which would make sense - eventually, one would expect the effect to be somewhat similar to neuron death or divergence with training gone wrong or just overtraining.

However, it can happen even with just one upvote or downvote image too.
Generating upvote/downvote images with one model, then applying them to another model seems to make it happen less.
Absolutely worst cases are >3 images generated with the same model and prompt they are applied to.

What am i doing wrong / is there an expected procedure to this?
Or is there a bug somewhere / lack of normalization of some kind leading to unet weights blowing up?

**Comment by dvruette** [comment]
I’ve run into similar problems myself before, but it’s kind of difficult to reproduce. Do you have some example feedback for which this reliably happens? Or some generation parameters to reproduce the behavior?

**Comment by orcinus** [comment]
Unfortunately, no, that's part of the problem - it appears very randomly.
I'll try and create some kind of a repro case.

**Comment by dvruette** [comment]
Yes, on one hand that sounds plausible, but on the other hand I’ve tested the plugin by replicating one image ~10 times without the model burning out. I suspect it has something to do with specific features in the image. For example dark patches with one solid color could be disproportionally influential on the generated image.

**Comment by orcinus** [comment]
Yeah, i had that thought too, but sometimes it's a dark or contrasty image that causes this, and sometimes a very non-contrasty flat one, without histogram extremes... I thought of testing it by manually tweaking feedback images, or even producing synthetic ones to see what happens, but didn't have time to do it yet.

**Comment by orcinus** [comment]
(WARNING: super long post)

I think i might be onto something...
I used this as a dumb, simple testbed for a synthetic feedback:

![FABRIC_orig](https://github.com/dvruette/sd-webui-fabric/assets/1574258/9c96919b-347a-484e-977b-78ba9b50a9b4)

I'm only using positive prompts, and only positive feedback to simplify things. The first thing to confirm was, how easy simple properties of the feedback image, like saturation, contrast and hue, transfer onto the generated images. So i took that image, created higher contrast version, lower contrast version, higher saturation version, lower saturation version, and shifted hue version and ran all of the variants with same seeds.

Original batch:

![FABRIC_result_orig](https://github.com/dvruette/sd-webui-fabric/assets/1574258/ba28731c-2f3f-4d5c-ab34-cbefa56daaba)

Lower contrast feedback:

![FABRIC_result_decontrast](https://github.com/dvruette/sd-webui-fabric/assets/1574258/d85fcaeb-db7b-47d4-b58f-dd6ebca15aed)

Higher contrast feedback:

![FABRIC_result_contrast](https://github.com/dvruette/sd-webui-fabric/assets/1574258/61e8a6e9-a527-4007-85bc-888d969cdb4e)

Lower saturation:

![FABRIC_result_desaturate](https://github.com/dvruette/sd-webui-fabric/assets/1574258/dc1ce0a1-fde2-423d-91f5-7afc2ad51b3d)


Higher saturation:

![FABRIC_result_saturate](https://github.com/dvruette/sd-webui-fabric/assets/1574258/1653aba1-12bb-43fd-abf2-720ee238fe3e)

Hue shift:

![FABRIC_result_hueshift](https://github.com/dvruette/sd-webui-fabric/assets/1574258/c1036f86-f423-4ad5-bcc8-c1eef546e545)

The conclusion - changes in hue, contrast, saturation VERY easily transfer from feedback images to generated images. And higher values seem to be especially sticky. So it's kind of obvious now why sometimes generations drift off into full on burnout.

Next thing i wanted to test is - how much does similarity (similar content) between the feedback image and images generated by the prompt affect, so i fabricated a feedback image that's just a circle on a background using the same dominant colors. Kind of an extreme simplification of the images generated:

![FABRIC_fabricated1](https://github.com/dvruette/sd-webui-fabric/assets/1574258/f3fce5fb-4fbd-4f37-92a5-5ce8ff55fd7e)

This is the result with it as feedback:

 
![fabricated_result_original](https://github.com/dvruette/sd-webui-fabric/assets/1574258/15d61e3e-3196-4b9e-ace7-5d78aa561726)

Lowered contrast:

![fabricated_result_decontrast](https://github.com/dvruette/sd-webui-fabric/assets/1574258/cae32ffd-6259-452e-901b-6173a100ec72)

Increased contrast:

![fabricated_result_contrast](https://github.com/dvruette/sd-webui-fabric/assets/1574258/deb27d49-49d6-4531-9917-98752bd70d89)

Desaturated:

![fabricated_result_desaturated](https://github.com/dvruette/sd-webui-fabric/assets/1574258/db9c3991-6c41-4da3-bf1f-fa6806d579f9)

Saturated:

![fabricated_result_saturated](https://github.com/dvruette/sd-webui-fabric/assets/1574258/720fb7aa-34c9-4bb6-9240-e71789b8e4c3)

Hue:

![fabricated_result_hue](https://github.com/dvruette/sd-webui-fabric/assets/1574258/7875a417-8c9b-4f05-99a4-6d88f1404734)

Sort of clearly evident most of the effect is on the sky and grass - i.e. areas that match by complexity / surface area.
So, in conclusion - global properties like contrast, hue, saturation, etc. easily transfer to the generated images, and the effect is progressively stronger the higher the similarity between the subjects or elements.

What to do about it?
No clue at the moment...

**Comment by orcinus** [comment]
Another unrelated thing i've noticed by accident, that i cannot quite understand...
Using images generated by model A as feedback for model B often times (always?) results in glitches and complete gibberish.
Talking "related" models here, so e.g. SD1.5 and fine tuned SD1.5.

**Comment by dvruette** [comment]
Very interesting, thanks for putting this together! I’ve definitely seen even more extreme examples where patches of the image are completely black or where the brightness is so low that the entire image is essentially black. I haven’t really thought about it in terms of contrast.

As far as finding a solution goes, I think it will require looking for a pattern in the model activations that’s different between burned-out images and normal images. It might also help finding more extreme examples because there it might be more obvious where exactly it goes wrong.

I’ve also never noticed that using different models for generating feedback results in glitches. Do you have some examples for that? It might well be related to the burnout issue.

**Comment by orcinus** [comment]
Yeah, these were very mild examples, but also, it didn't take a lot to produce them.
I've seen complete burnouts too.

I'll try and make different model glitch repro case too.

What i *think* might maybe help is bounding the maximum weights of the feedback to some range.
Question is what that range should be... i think i've seen a similar issue way back a year or so ago in the early days of SD when someone was trying to prevent the image "burn" at high CFGs. It turned out that it's not as easy as it initially sounds, because there's no good way to map latent space value ranges to image value ranges (i.e. 0-255) across different generations.

**Comment by dvruette** [comment]
I've been able to find a pair of feedback images that reliably produces burnout, so I'm posting them here for future reference. Some observations I've been able to make so far:
- Larger CFG scales produce more burnout
- Bunout heavily depends on the prompt
  - the prompt used here is `award winning oil painting of a corgi sitting in a garden, soothing tones, colorful, high contrast, (masterpiece), (best quality:1.2), absurdres, intricate details, wide angle, in the style of monet`
- The effect can be somewhat diminished by using an appropriate negative prompt, e.g. `black background`
- Either one of the feedback images is enough to produce burnout, but combining both amplifies the effect.

Feedback image 1:
![00063-948568555](https://github.com/dvruette/sd-webui-fabric/assets/12776676/3b6bd1c7-3e86-40e3-8cf8-d148ce9d3214)

Feedback image 2:
![00074-948568555](https://github.com/dvruette/sd-webui-fabric/assets/12776676/b7dca041-998d-4b8c-b96d-8d37db003506)

Example generation with/without feedback:
![grid-0031](https://github.com/dvruette/sd-webui-fabric/assets/12776676/99fc672c-704c-481c-aba9-ebaf35d82202)
![grid-0030](https://github.com/dvruette/sd-webui-fabric/assets/12776676/66a1b8cb-c524-4a8f-ba7b-f50a94e81ca4)

**Comment by orcinus** [comment]
Interesting...
The pattern visible in the worst examples is similar to what you sometimes get with img2img when accidentally using the same seed on both the original image and the current generation.

**Comment by dvruette** [comment]
Oh interesting, this might be an indication that predicted noise updates are too large or otherwise OOD. Not very familiar with the img2img failure mode, but do you know if adaptive CFG scale and/or more denoising steps help at all?

**Comment by orcinus** [comment]
They usually just make the pattern less/more obvious.
I have no idea what causes it, but i *think* it's probably reinforcement of existing noise in *some* way.

As in, any leftover noise is very similar and gets signal-boosted through repetition.

**Comment by dvruette** [comment]
It's been a while, but I've finally found a way to mitigate this issue and mostly prevent it from happening.
